File: src/main/java/jp/kentan/minecraft/neko_event/config/provider/ParkourConfigProvider.java
package jp.kentan.minecraft.neko_event.config.provider;

import jp.kentan.minecraft.neko_event.config.ConfigManager;
import jp.kentan.minecraft.neko_event.config.listener.ConfigListener;
import jp.kentan.minecraft.neko_event.game.parkour.model.Parkour;
import jp.kentan.minecraft.neko_event.util.Log;
import org.apache.commons.lang.time.DateUtils;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.configuration.file.YamlConfiguration;

import java.io.File;
import java.util.*;

import static jp.kentan.minecraft.neko_event.util.NekoUtil.DATE_FORMAT;

public class ParkourConfigProvider {

    private static File sConfigFile;

    private static ConfigListener<Parkour> sListener;

    public static void setup(File dataFolder){
        sConfigFile = new File(dataFolder + File.separator + "parkour.yml");

        ConfigManager.createIfNeed(sConfigFile);
    }

    public static void bindListener(ConfigListener<Parkour> listener){
        sListener = listener;
    }

    public static void load() {
        try {
            FileConfiguration config = new YamlConfiguration();
            config.load(sConfigFile);

            final Map<String, Parkour> parkourMap = new HashMap<>();

            if(!config.isConfigurationSection("Parkour")){
                Log.warn("Parkour section not found.");
                return;
            }

            final Set<String> parkourIdSet = config.getConfigurationSection("Parkour").getKeys(false);

            parkourIdSet.forEach(id -> {
                final String path = "Parkour." + id;

                Parkour parkour = new Parkour(id,
                        config.getString(path + ".name", "アスレチック"),
                        config.isConfigurationSection(path + ".Join.Location") ?
                                new Location(
                                        Bukkit.getWorld(config.getString(path + ".Join.Location.world")),
                                        config.getDouble(path + ".Join.Location.x"),
                                        config.getDouble(path + ".Join.Location.y"),
                                        config.getDouble(path + ".Join.Location.z"),
                                        (float)config.getDouble(path + ".Join.Location.yaw"),
                                        (float)config.getDouble(path + ".Join.Location.pitch")
                                ) : null,
                        config.isConfigurationSection(path + ".Clear.Location") ?
                                new Location(
                                        Bukkit.getWorld(config.getString(path + ".Clear.Location.world")),
                                        config.getDouble(path + ".Clear.Location.x"),
                                        config.getDouble(path + ".Clear.Location.y"),
                                        config.getDouble(path + ".Clear.Location.z"),
                                        (float)config.getDouble(path + ".Clear.Location.yaw"),
                                        (float)config.getDouble(path + ".Clear.Location.pitch")
                                ) : null,
                        config.isConfigurationSection(path + ".Back.Location") ?
                                new Location(
                                        Bukkit.getWorld(config.getString(path + ".Back.Location.world")),
                                        config.getDouble(path + ".Back.Location.x"),
                                        config.getDouble(path + ".Back.Location.y"),
                                        config.getDouble(path + ".Back.Location.z"),
                                        (float)config.getDouble(path + ".Back.Location.yaw"),
                                        (float)config.getDouble(path + ".Back.Location.pitch")
                                ) : null,
                        config.getString(path + ".Join.playerMsg"),
                        config.getString(path + ".Join.broadcastMsg"),
                        config.getString(path + ".Clear.playerMsg"),
                        config.getString(path + ".Clear.broadcastMsg"),
                        config.getString(path + ".Clear.titleText"),
                        config.getString(path + ".Back.playerMsg")
                );

                parkourMap.put(id, parkour);
            });

            Log.info(parkourMap.size() + "個のアスレチックを読み込みました.");

            sListener.onUpdate(parkourMap);
        } catch (Exception e) {
            e.printStackTrace();
            Log.error(e.getMessage());
        }
    }

    public static Parkour create(String id, String name, Location joinLocation, Location clearLocation, Location backLocation, String joinPlayerMsg,
                                 String joinBroadcastMsg, String clearPlayerMsg, String clearBroadcastMsg, String clearTitleText, String baclPlayerMsg) {
        final String path = "Parkour." + id;

        try {
            FileConfiguration config = new YamlConfiguration();
            config.load(sConfigFile);

            if(config.isConfigurationSection(path)){
                Log.error("パルクール({id})は既に登録されています.".replace("{id}",id));
                return null;
            }
        } catch (Exception e) {
            e.printStackTrace();
            Log.error(e.getMessage());
            return null;
        }

        if(!save(new LinkedHashMap<String, Object>(){
            {
                put(path + ".name", name);

                if(joinLocation != null) {
                    put(path + ".Join.Location.world", joinLocation.getWorld().getName());
                    put(path + ".Join.Location.x", joinLocation.getX());
                    put(path + ".Join.Location.y", joinLocation.getY());
                    put(path + ".Join.Location.z", joinLocation.getZ());
                    put(path + ".Join.Location.yaw", (double) joinLocation.getYaw());
                    put(path + ".Join.Location.pitch", (double) joinLocation.getPitch());
                }

                if(clearLocation != null) {
                    put(path + ".Clear.Location.world", clearLocation.getWorld().getName());
                    put(path + ".Clear.Location.x", clearLocation.getX());
                    put(path + ".Clear.Location.y", clearLocation.getY());
                    put(path + ".Clear.Location.z", clearLocation.getZ());
                    put(path + ".Clear.Location.yaw", (double) clearLocation.getYaw());
                    put(path + ".Clear.Location.pitch", (double) clearLocation.getPitch());
                }

                if(backLocation != null) {
                    put(path + ".Back.Location.world", backLocation.getWorld().getName());
                    put(path + ".Back.Location.x", backLocation.getX());
                    put(path + ".Back.Location.y", backLocation.getY());
                    put(path + ".Back.Location.z", backLocation.getZ());
                    put(path + ".Back.Location.yaw", (double) backLocation.getYaw());
                    put(path + ".Back.Location.pitch", (double) backLocation.getPitch());
                }

                put(path + ".Join.playerMsg", joinPlayerMsg);
                put(path + ".Join.broadcastMsg", joinBroadcastMsg);
                put(path + ".Clear.playerMsg", clearPlayerMsg);
                put(path + ".Clear.broadcastMsg", clearBroadcastMsg);
                put(path + ".Clear.titleText", clearTitleText);
                put(path + ".Back.playerMsg", baclPlayerMsg);
            }
        })){
            return null;
        }

        return new Parkour(id, name, joinLocation, clearLocation, backLocation, joinPlayerMsg, joinBroadcastMsg, clearPlayerMsg, clearBroadcastMsg, clearTitleText, baclPlayerMsg);
    }

    public static boolean delete(String id) {
        final String path = "Parkour." + id;

        return save(new HashMap<String, Object>(){
            {
                put(path , null);
            }
        });
    }

    public static boolean update(String id, Location joinLocation, Location clearLocation, Location backLocation, String joinPlayerMsg, String joinBroadcastMsg, String clearPlayerMsg,
                                 String clearBroadcastMsg, String clearTitleText, String backPlayerMsg) {
        final String path = "Parkour." + id;

        try {
            FileConfiguration config = new YamlConfiguration();
            config.load(sConfigFile);

            if(!config.isConfigurationSection(path)){
                Log.error("パルクール({id})は存在しません.".replace("{id}",id));
                return false;
            }
        } catch (Exception e) {
            e.printStackTrace();
            Log.error(e.getMessage());
            return false;
        }

        return save(new LinkedHashMap<String, Object>(){
            {
                if(joinLocation != null) {
                    put(path + ".Join.Location.world", joinLocation.getWorld().getName());
                    put(path + ".Join.Location.x", joinLocation.getX());
                    put(path + ".Join.Location.y", joinLocation.getY());
                    put(path + ".Join.Location.z", joinLocation.getZ());
                    put(path + ".Join.Location.yaw", (double) joinLocation.getYaw());
                    put(path + ".Join.Location.pitch", (double) joinLocation.getPitch());
                }

                if(clearLocation != null){
                    put(path + ".Clear.Location.world", clearLocation.getWorld().getName());
                    put(path + ".Clear.Location.x", clearLocation.getX());
                    put(path + ".Clear.Location.y", clearLocation.getY());
                    put(path + ".Clear.Location.z", clearLocation.getZ());
                    put(path + ".Clear.Location.yaw", (double)clearLocation.getYaw());
                    put(path + ".Clear.Location.pitch", (double)clearLocation.getPitch());
                }

                if(backLocation != null){
                    put(path + ".Back.Location.world", backLocation.getWorld().getName());
                    put(path + ".Back.Location.x", backLocation.getX());
                    put(path + ".Back.Location.y", backLocation.getY());
                    put(path + ".Back.Location.z", backLocation.getZ());
                    put(path + ".Back.Location.yaw", (double)backLocation.getYaw());
                    put(path + ".Back.Location.pitch", (double)backLocation.getPitch());
                }

                if(joinPlayerMsg != null) {
                    put(path + ".Join.playerMsg", joinPlayerMsg);
                }
                if(joinBroadcastMsg != null) {
                    put(path + ".Join.broadcastMsg", joinBroadcastMsg);
                }

                if(clearPlayerMsg != null) {
                    put(path + ".Clear.playerMsg", clearPlayerMsg);
                }
                if(clearBroadcastMsg != null) {
                    put(path + ".Clear.broadcastMsg", clearBroadcastMsg);
                }
                if(clearTitleText != null){
                    put(path + ".Clear.titleText", clearTitleText);
                }

                if(backPlayerMsg != null) {
                    put(path + ".Back.playerMsg", backPlayerMsg);
                }
            }
        });
    }

    public static boolean saveClearDate(UUID uuid, String id){
        return PlayerConfigProvider.save(uuid, new HashMap<String, Object>(){
            {
                put("Parkour." + id + ".clearDate", DATE_FORMAT.format(new Date()));
            }
        });
    }

    public static boolean hasClearedToday(UUID uuid, String id){
        try {
            final String strPrev = (String) PlayerConfigProvider.get(uuid, "Parkour." + id + ".clearDate", null);

            if(strPrev == null){
                return false;
            }

            final Date prev = DATE_FORMAT.parse(strPrev);
            final Date today = DateUtils.truncate(new Date(), Calendar.DAY_OF_MONTH);

            return prev != null && prev.compareTo(today) == 0;
        }catch (Exception e){
            e.printStackTrace();
            Log.error(e.getMessage());
            return false;
        }
    }

    private static boolean save(Map<String, Object> dataList) {
        try {
            FileConfiguration config = new YamlConfiguration();
            config.load(sConfigFile);

            dataList.forEach(config::set);

            config.save(sConfigFile);
        } catch (Exception e) {
            e.printStackTrace();
            Log.error(e.getMessage());

            return false;
        }
        return true;
    }
}

File: src/main/java/jp/kentan/minecraft/nekoevent/config/provider/ParkourConfigProvider.kt
package jp.kentan.minecraft.nekoevent.config.provider

import jp.kentan.minecraft.nekoevent.component.model.Parkour
import jp.kentan.minecraft.nekoevent.config.ConfigManager.Companion.DATE_FORMAT
import jp.kentan.minecraft.nekoevent.config.ConfigUpdateListener
import jp.kentan.minecraft.nekoevent.util.Log
import org.apache.commons.lang.time.DateUtils
import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.configuration.file.YamlConfiguration
import org.bukkit.entity.Player
import java.io.File
import java.util.*

class ParkourConfigProvider(
        dataFolder: File,
        private val playerConfig: PlayerConfigProvider
) : BaseConfigProvider(dataFolder, "parkour.yml") {

    var listener: ConfigUpdateListener<Parkour>? = null

    fun load() {
        try {
            val config = YamlConfiguration.loadConfiguration(configFile)

            if (!config.isConfigurationSection("Parkour")) {
                Log.warn("Parkourセクションがありません.")
                return
            }

            val parkourIdSet = config.getConfigurationSection("Parkour")?.getKeys(false).orEmpty()
            val parkourMap = parkourIdSet.associate { id ->
                val path = "Parkour.$id"

                val parkour = Parkour(
                        id,
                        config.getString("$path.name") ?: "パルクール",
                        config.getInt("$path.Reward.ticketAmount"),
                        if (config.isConfigurationSection("$path.Join.Location"))
                            Location(
                                    Bukkit.getWorld(config.getString("$path.Join.Location.world").orEmpty()),
                                    config.getDouble("$path.Join.Location.x"),
                                    config.getDouble("$path.Join.Location.y"),
                                    config.getDouble("$path.Join.Location.z"),
                                    config.getDouble("$path.Join.Location.yaw").toFloat(),
                                    config.getDouble("$path.Join.Location.pitch").toFloat()
                            )
                        else
                            null,
                        config.getString("$path.Join.message"),
                        config.getString("$path.Join.broadcastMessage"),
                        if (config.isConfigurationSection("$path.Clear.Location"))
                            Location(
                                    Bukkit.getWorld(config.getString("$path.Clear.Location.world").orEmpty()),
                                    config.getDouble("$path.Clear.Location.x"),
                                    config.getDouble("$path.Clear.Location.y"),
                                    config.getDouble("$path.Clear.Location.z"),
                                    config.getDouble("$path.Clear.Location.yaw").toFloat(),
                                    config.getDouble("$path.Clear.Location.pitch").toFloat()
                            )
                        else
                            null,
                        config.getString("$path.Clear.message"),
                        config.getString("$path.Clear.broadcastMessage"),
                        if (config.isConfigurationSection("$path.Back.Location"))
                            Location(
                                    Bukkit.getWorld(config.getString("$path.Back.Location.world").orEmpty()),
                                    config.getDouble("$path.Back.Location.x"),
                                    config.getDouble("$path.Back.Location.y"),
                                    config.getDouble("$path.Back.Location.z"),
                                    config.getDouble("$path.Back.Location.pitch").toFloat(),
                                    config.getDouble("$path.Back.Location.yaw").toFloat()
                            )
                        else
                            null,
                        config.getString("$path.Back.message")
                )

                return@associate id to parkour
            }

            listener?.onConfigUpdate(parkourMap)
        } catch (e: Exception) {
            e.printStackTrace()
            Log.error(e)
        }
    }

    fun update(parkour: Parkour): Boolean {
        val path = "Parkour.${parkour.id}"
        val dataMap = LinkedHashMap<String, Any?>().apply {
            put("$path.name", parkour.name)

            put("$path.Reward.ticketAmount", parkour.rewardTicketAmount)

            if (parkour.joinLocation != null) {
                val loc = parkour.joinLocation
                put("$path.Join.Location.world", loc.world?.name)
                put("$path.Join.Location.x", loc.x)
                put("$path.Join.Location.y", loc.y)
                put("$path.Join.Location.z", loc.z)
                put("$path.Join.Location.yaw", loc.yaw)
                put("$path.Join.Location.pitch", loc.pitch)
            } else {
                put("$path.Join.Location", null)
            }
            put("$path.Join.message", parkour.joinMessage)
            put("$path.Join.broadcastMessage", parkour.joinBroadcastMessage)

            if (parkour.clearLocation != null) {
                val loc = parkour.clearLocation
                put("$path.Clear.Location.world", loc.world?.name)
                put("$path.Clear.Location.x", loc.x)
                put("$path.Clear.Location.y", loc.y)
                put("$path.Clear.Location.z", loc.z)
                put("$path.Clear.Location.yaw", loc.yaw)
                put("$path.Clear.Location.pitch", loc.pitch)
            } else {
                put("$path.Clear.Location", null)
            }
            put("$path.Clear.message", parkour.clearMessage)
            put("$path.Clear.broadcastMessage", parkour.clearBroadcastMessage)

            if (parkour.backLocation != null) {
                val loc = parkour.backLocation
                put("$path.Back.Location.world", loc.world?.name)
                put("$path.Back.Location.x", loc.x)
                put("$path.Back.Location.y", loc.y)
                put("$path.Back.Location.z", loc.z)
                put("$path.Back.Location.yaw", loc.yaw)
                put("$path.Back.Location.pitch", loc.pitch)
            } else {
                put("$path.Back.Location", null)
            }
            put("$path.Back.message", parkour.backMessage)
        }

        return super.save(dataMap)
    }

    fun delete(keyId: String): Boolean {
        return super.save(mapOf("Parkour.$keyId" to null))
    }

    fun hasClearedToday(player: Player, parkour: Parkour): Boolean {
        try {
            val oldDate = playerConfig.get(player.uniqueId, "Parkour.${parkour.id}.clearDate") as String? ?: return false

            val prev = DATE_FORMAT.parse(oldDate)
            val today = DateUtils.truncate(Date(), Calendar.DAY_OF_MONTH)

            return prev.compareTo(today) == 0
        } catch (e: Exception) {
            e.printStackTrace()
            Log.error(e)
        }

        return false
    }

    fun saveClearDate(player: Player, parkour: Parkour) {
        playerConfig.save(player.uniqueId, mapOf("Parkour.${parkour.id}.clearDate" to DATE_FORMAT.format(Date())))
    }
}